File: java/firstmissingpositive.java
public class Solution {
    public int firstMissingPositive(int[] A) {
        // Start typing your Java solution below
        // DO NOT write main() function
        for(int i=0;i<A.length;i++){
            if(A[i]<=0) A[i] = Integer.MAX_VALUE;
        }
        for(int i=0;i<A.length;i++){
            if(A[i]>0 && A[i]-1 < A.length && A[A[i]-1]!=A[i] ){
                int k=A[i];
                A[i]=A[k-1];
                A[k-1]=k;
                i--;
            }
        }
        for(int i=1;i<=A.length;i++){
            if(A[i-1]!=i){
                return i;
            }
        }
        return A.length+1;
    }
}

File: cpp/FirstMissingPositive.cpp
class Solution {
public:
    int firstMissingPositive(int A[], int n) {
        int max = 0;
        for(int i=0;i<n;i++) {
            max = (A[i]>max)? A[i]:max;
        }
        if(max==0) {
            return 1;
        }
        for(int i=0;i<n;i++) {
            if(A[i] <=0) {
                A[i] = max+1;
            }
        }
        int t = 0;
        for(int i=0;i<n;i++) {
            if(abs(A[i])==max+1) {
                continue;
            } else {
                t = abs(A[i])-1;
                if(t<n) {
                    A[t] = -abs(A[t]);
                }
            }
        }
        for(int i=0;i<n;i++) {
            if(A[i]>0) {
                return i+1;
            }
        }
        return n+1;
    }
};
